db/schema_tables.cc

```diff
@@ -974,23 +990,32 @@ static read_table_names_of_keyspace(distributed<service::storage_proxy>& proxy,
     }));
 }
 
-static table_id table_id_from_mutations(const schema_mutations& sm) {
+static std::optional<table_id> table_id_from_mutations(const schema_mutations& sm) {
     auto table_rs = query::result_set(sm.columnfamilies_mutation());
+    if (table_rs.empty()) {
+        return std::nullopt;
+    }
     query::result_set_row table_row = table_rs.row(0);
     return table_id(table_row.get_nonnull<utils::UUID>("id"));
 }
 
 static
 future<std::map<table_id, schema_mutations>>
-read_tables_for_keyspaces(distributed<service::storage_proxy>& proxy, const std::set<sstring>& keyspace_names, schema_ptr s)
+read_tables_for_keyspaces(distributed<service::storage_proxy>& proxy, const std::set<sstring>& keyspace_names, schema_ptr s,
+                          const std::unordered_map<sstring, std::optional<std::set<sstring>>>& tables_per_keyspace)
```

Comment:
Long-standing dislike about the fact that we read anything from disk. We have the entire schema in memory already.

Reply:
Why `std::optional<std::set>`? Won't an empty set, or missing entry in the map suffice?

Reply:
missing entry means we haven't proccessed anything for the keyspace yet, while std::nullopt means that we processed some mutation and figured out that the set of affected tables is potentially unbounded. In this case, we must not override std::nullopt with a more narrow set coming from another mutation. An empty set is no good too, because it means no tables are affected, while std::nullopt means that many are affected.

Reply:
Yes, I saw it, but it's adding meaning to std::optional in tricky and rarely touched code. We should be more explicit.

A counterexample is std::optional used for NULL, but there the meaning is similar, and the code is widely used so it's reasonable to assume the reader is familiar with the convention.